packages/firestore/test/integration/remote/stream.test.ts

```diff
@@ -0,0 +1,231 @@
+  private pendingStates: StreamStatusCallback[] = [];
+  private pendingPromises: Deferred<StreamStatusCallback>[] = [];
+
+  private resolvePending(actualCallback: StreamStatusCallback) {
+    let pendingPromise = this.pendingPromises.shift();
+    if (pendingPromise) {
+      pendingPromise.resolve(actualCallback);
+    } else {
+      this.pendingStates.push(actualCallback);
+    }
+    return Promise.resolve();
+  }
+
+  onHandshakeComplete(): Promise<void> {
+    return this.resolvePending('onHandshakeComplete');
+  }
+
+  onMutationResult(
+    commitVersion: SnapshotVersion,
+    results: MutationResult[]
+  ): Promise<void> {
+    return this.resolvePending('onMutationResult');
+  }
+
+  onWatchChange(
+    watchChange:
+      | DocumentWatchChange
+      | WatchTargetChange
+      | ExistenceFilterChange,
+    snapshot: SnapshotVersion
+  ): Promise<void> {
+    return this.resolvePending('onWatchChange');
+  }
+
+  onOpen(): Promise<void> {
+    return this.resolvePending('onOpen');
+  }
+
+  onClose(err?: FirestoreError): Promise<void> {
+    return this.resolvePending('onClose');
+  }
+
+  /**
+   * Returns a Promise that resolves when 'expectedCallback' fires.
+   * Resolves the returned Promise immediately if there is already an
+   * unprocessed callback.
+   * Fails the test if the expected callback does not match the name of the
+   * actual callback function.
+   */
+  awaitCallback(expectedCallback: StreamStatusCallback): Promise<void> {
+    if (this.pendingStates.length > 0) {
+      expect(this.pendingStates.shift()).to.equal(expectedCallback);
+      return Promise.resolve();
+    } else {
+      const deferred = new Deferred<StreamStatusCallback>();
+      this.pendingPromises.push(deferred);
+      return deferred.promise.then(actualCallback => {
+        expect(actualCallback).to.equal(expectedCallback);
+      });
+    }
+  }
+
+  verifyNoPendingCallbacks(): void {
+    expect(this.pendingStates).to.be.empty;
+  }
+}
+
+describe('Watch Stream', () => {
+  let queue: AsyncQueue;
+  let streamListener: StreamStatusListener;
+
+  beforeEach(() => {
+    queue = new AsyncQueue();
+    streamListener = new StreamStatusListener();
+  });
+
+  afterEach(() => {
+    streamListener.verifyNoPendingCallbacks();
+  });
+
+  function initializeWatchStream(): Promise<PersistentListenStream> {
+    const databaseInfo = getDefaultDatabaseInfo();
+
+    return PlatformSupport.getPlatform()
+      .loadConnection(databaseInfo)
+      .then(conn => {
+        const serializer = PlatformSupport.getPlatform().newSerializer(
+          databaseInfo.databaseId
+        );
+        return new Datastore(
+          databaseInfo,
+          queue,
+          conn,
+          new EmptyCredentialsProvider(),
+          serializer
+        );
+      })
+      .then(ds => {
+        return ds.newPersistentWatchStream(streamListener);
+      });
+  }
+
+  asyncIt('can be stopped before handshake', () => {
```

Comment:
The iOS test has a comment:

/** Verifies that the watch stream does not issue an onClose callback after a call to stop(). */

and does indeed trigger a GRPC close event (to make sure that our stream doesn't trigger a close event in response). Why aren't we testing that on web?

Reply:
I went back and forth on this just now, but decided to ultimately leave it as is. The handleStreamClose callback on the Web only ever gets called when the stream is still open and asserts that this is the case. To bring it closer to the other platforms, I would have to break this invariant.

If you prefer, I can also assert here that 'handleStreamClose' throws an error at this point. Just making it public for that doesn't seem worth it - especially since it invites other consumers to trip over this assertion.

Reply:
I see.  To make web a more direct port, I think we'd refactor https://github.com/firebase/firebase-js-sdk/blob/fba3c06cfd7baee797c1bec629ade4c200b01d18/packages/firestore/src/remote/persistent_stream.ts#L339 so that it delegates to a function that could then be called from this test.  But going to that much trouble probably isn't warranted.